가중치의 초기값은 0으로 시작하고, 학습률은 0.1로 설정하겠습니다.

퍼셉트론의 출력 함수는 입력과 가중치의 합이 0보다 크거나 같으면 1, 그렇지 않으면 0으로 정의하겠습니다.

가중치 업데이트는 아래와 같은 공식에 따라 진행됩니다: w = w + η * (d - y) * x


다음 용어를 설명하라


(1) 하이퍼파라미터(Hyperparameter):
하이퍼파라미터란 머신러닝 모델의 구조나 학습 알고리즘에 영향을 주는 매개변수를 의미합니다. 이러한 매개변수는 모델 학습 전에 사람이 수동으로 설정해야 하며, 학습 과정 중에는 자동으로 조정되지 않습니다. 예를 들어, 신경망 모델에서 은닉층의 수, 각 층의 뉴런 개수, 학습률 등이 하이퍼파라미터입니다. 이러한 하이퍼파라미터를 조정하면 모델의 성능과 학습 속도에 영향을 줄 수 있습니다.

(2) 기울기 소실(Vanishing Gradient):
기울기 소실은 신경망 학습 과정에서 발생하는 문제로, 역전파 알고리즘을 사용하여 신경망을 학습시킬 때 일어납니다. 신경망의 깊이가 깊어질수록, 역전파 알고리즘에 의해 그래디언트(기울기)가 출력층에서 입력층으로 전달되면서 작아지는 현상입니다. 이러한 작은 그래디언트는 가중치 업데이트에 거의 영향을 주지 않아 신경망이 제대로 학습되지 않고, 성능이 좋지 않게 될 수 있습니다.

(3) 배치 정규화(Batch Normalization):
배치 정규화는 딥러닝 모델에서 사용되는 기법 중 하나로, 학습 과정 동안 각 미니배치의 입력 데이터를 정규화하는 과정입니다. 배치 정규화는 신경망의 각 층에서 활성화 값을 조정하여 그래디언트의 사라지는 문제를 완화하고, 학습 속도를 향상시키는 효과가 있습니다. 또한, 가중치 초기화에 대한 의존성을 줄이고, 모델의 일반화 성능을 향상시킵니다.

(4) 옵티마이저(Optimizer):
옵티마이저는 머신러닝 모델의 학습 과정에서 손실 함수를 최적화하는 방법을 결정하는 알고리즘입니다. 옵티마이저는 모델의 가중치와 편향을 업데이트하여 손실 함수를 최소화하도록 합니다. 예를 들어, 경사 하강법(Gradient Descent) 알고리즘은 옵티마이저의 일종으로, 그래디언트를 사용하여 손실 함수를 최소화하는 방향으로 모델을 업데이트합니다. 다른 옵티마이저에는 모멘텀 최적화(Momentum Optimization), 아담(Adam) 등이 있습니다.

(5) 드롭아웃(Dropout):
드롭아웃은 신경망에서 사용되는 정규화 기법 중 하나로, 학습 과정 중에 랜덤하게 일부 뉴런을 제거하고, 제거된 뉴런을 포함한 신경망으로 학습을 진행하는 방법입니다. 이를 통해 모델이 특정 뉴런에 지나치게 의존하지 않도록 하여 과적합을 방지하고, 모델의 일반화 능력을 향상시킬 수 있습니다.

(6) 합성곱(Convolution):
합성곱은 신경망에서 이미지 처리와 같은 공간 정보를 가진 데이터를 처리하기 위한 연산입니다. 합성곱은 커널(필터)과 입력 데이터의 원소간의 곱셈과 합을 통해 새로운 특성 맵을 생성합니다. 이를 통해 이미지의 공간적인 패턴을 감지하거나 특성을 추출하는데 사용됩니다. 합성곱 연산은 합성곱 신경망(Convolutional Neural Network, CNN)의 주요 구성 요소로 활용되며, 이미지 분류, 객체 검출 등에 널리 사용됩니다.

(7) 전이 학습(Transfer Learning):
전이 학습은 한 도메인에서 학습된 모델의 지식을 다른 도메인의 학습에 활용하는 방법입니다. 전이 학습은 기존에 큰 규모의 데이터셋에서 학습된 모델을 가져와서 새로운 작업에 활용하거나, 비슷한 도메인에서 학습된 모델의 일부를 재사용하여 학습 속도와 성능을 개선하는 데 사용됩니다. 이를 통해 적은 데이터로도 효과적인 모델을 학습할 수 있으며, 실제 응용 분야에서 많이 활용되는 기법입니다.

(8) 특이값 분해(Singular Value Decomposition, SVD):
특이값 분해는 선형 대수학에서 사용되는 기법으로, 행렬을 세 개의 행렬의 곱으로 분해하는 방법입니다. 주어진 행렬을 세 가지 행렬의 곱으로 분해함으로써, 행렬의 특성을 파악하고 정보를 추출할 수 있습니다. 특이값 분해는 차원 축소, 행렬의 랭크 결정, 행렬의 역행렬 계산 등 다양한 분야에서 사용됩니다.

(9) 매니폴드(Manifold):
매니폴드는 고차원 공간에서의 저차원 부분 공간을 나타내는 수학적인 개념입니다. 고차원 데이터는 저차원 매니폴드 상에 분포되어 있을 때, 이를 이해하고 처리하는 데 유용합니다. 매니폴드 학습은 고차원 데이터를 저차원의 매니폴드 공간으로 변환하여 데이터의 차원을 축소하고, 주요한 정보를 추출하는 방법입니다. 이를 통해 데이터 시각화, 차원 축소, 패턴 인식 등에 활용됩니다.

(10) 잠재 표현(Latent Representation):
잠재 표현은 머신러닝 모델에서 입력 데이터를 나타내는 고차원 벡터로, 데이터의 중요한 특성을 포착하고 나타냅니다. 이러한 잠재 표현은 데이터의 특성을 나타내는 요소들로 구성되며, 데이터를 표현하거나 변환하는 데 사용됩니다. 잠재 표현은 차원 축소, 군집화, 생성 모델 등 다양한 응용 분야에서 활용됩니다.


1. Forward Propagation (순전파):
- 입력층에서 은닉층으로의 전달: 입력값과 가중치를 곱하고, 활성화 함수(sigmoid)를 적용하여 은닉층의 출력을 계산합니다.
- 은닉층에서 출력층으로의 전달: 은닉층의 출력과 은닉층과 출력층 사이의 가중치를 곱하고, 활성화 함수(sigmoid)를 적용하여 출력층의 출력을 계산합니다.


2. 계산된 출력과 실제 출력을 비교하여 오차를 계산합니다.


3. Backward Propagation (오류역전파):
- 출력층에서 은닉층으로의 역전파:
  - 출력층의 오차를 계산하여 은닉층과 출력층 사이의 가중치(w)를 수정합니다.
  - 은닉층의 출력과 가중치 w의 변화율을 곱하여 은닉층의 오차를 계산합니다.
- 은닉층에서 입력층으로의 역전파:
  - 은닉층의 오차를 계산하여 입력층과 은닉층 사이의 가중치(u1, u2)를 수정합니다.


4. 가중치 수정:
- 계산된 오차를 이용하여 가중치를 수정합니다. 일반적으로 경사하강법(Gradient Descent)을 사용하여 가중치를 업데이트합니다.
- 은닉층과 출력층 사이의 가중치 w를 수정하고, 입력층과 은닉층 사이의 가중치 u1, u2를 수정합니다.


5. 위 과정을 반복하면서 가중치를 업데이트하고, 오차를 최소화하는 방향으로 학습을 진행합니다. 학습을 여러 번 반복하면 네트워크가 데이터에 대해 더 나은 예측을 할 수 있도록 학습됩니다.

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

# Fashion-MNIST 데이터셋 로드
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# 데이터 전처리
train_images = train_images / 255.0
test_images = test_images / 255.0

# 데이터를 1차원 벡터로 변환
train_images = train_images.reshape(-1, 28*28)
test_images = test_images.reshape(-1, 28*28)

# MLP 모델 구성
model = keras.Sequential([
    keras.layers.Dense(256, activation='relu', input_shape=(28*28,)),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(10, activation='softmax')
])

# 모델 컴파일
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

# 모델 학습
model.fit(train_images, train_labels, epochs=50, batch_size=128, validation_split=0.3)

# 테스트 데이터로 모델 평가
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('테스트 정확도:', test_acc)


4422102/4422102 [==============================] - 0s 0us/step
Epoch 1/50


/usr/local/lib/python3.10/dist-packages/keras/backend.py:5612: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


329/329 [==============================] - 7s 5ms/step - loss: 0.7010 - accuracy: 0.7544 - val_loss: 0.4576 - val_accuracy: 0.8306
Epoch 2/50
329/329 [==============================] - 1s 4ms/step - loss: 0.4597 - accuracy: 0.8377 - val_loss: 0.3890 - val_accuracy: 0.8556
Epoch 3/50
329/329 [==============================] - 1s 4ms/step - loss: 0.4189 - accuracy: 0.8498 - val_loss: 0.3667 - val_accuracy: 0.8678
Epoch 4/50
329/329 [==============================] - 2s 6ms/step - loss: 0.3882 - accuracy: 0.8600 - val_loss: 0.3651 - val_accuracy: 0.8669
Epoch 5/50
329/329 [==============================] - 2s 5ms/step - loss: 0.3700 - accuracy: 0.8663 - val_loss: 0.3535 - val_accuracy: 0.8716
Epoch 6/50
329/329 [==============================] - 1s 4ms/step - loss: 0.3534 - accuracy: 0.8722 - val_loss: 0.3457 - val_accuracy: 0.8739
Epoch 7/50
329/329 [==============================] - 1s 4ms/step - loss: 0.3374 - accuracy: 0.8764 - val_loss: 0.3451 - val_accuracy: 0.8749
Epoch 8/50
329/32

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 256)               200960    
                                                                 
 dropout (Dropout)           (None, 256)               0         
                                                                 
 dense_1 (Dense)             (None, 64)                16448     
                                                                 
 dropout_1 (Dropout)         (None, 64)                0         
                                                                 
 dense_2 (Dense)             (None, 10)                650       
                                                                 
Total params: 218,058
Trainable params: 218,058
Non-trainable params: 0
_________________________________________________________________


첫 번째 은닉층:
  - 입력 크기: 784
  - 노드 수: 256
  - 파라미터 개수 = (입력 크기 + 1) * 노드 수 = (784 + 1) * 256 = 200,960


두 번째 은닉층:
  - 입력 크기: 256
  - 노드 수: 64
  - 파라미터 개수 = (입력 크기 + 1) * 노드 수 = (256 + 1) * 64 = 16,448


출력층:
  - 입력 크기: 64
  - 노드 수: 10
  - 파라미터 개수 = (입력 크기 + 1) * 노드 수 = (64 + 1) * 10 = 650

In [1]:
import tensorflow as tf
from tensorflow import keras

# Fashion-MNIST 데이터셋 로드
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# 데이터 전처리
train_images = train_images / 255.0
test_images = test_images / 255.0

# CNN 모델 구성
model = keras.Sequential([
    keras.layers.Conv2D(64, (3, 3), strides=(2, 2), padding='same', activation='relu', input_shape=(28, 28, 1)),
    keras.layers.MaxPooling2D((2, 2)),
    keras.layers.Conv2D(64, (3, 3), strides=2, padding='same', activation='relu'),
    keras.layers.MaxPooling2D((2, 2)),
    keras.layers.Flatten(),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(10, activation='softmax')
])

# 모델 컴파일
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=False),
              metrics=['accuracy'])

# 데이터 차원 변경 (1차원을 3차원으로)
train_images = train_images.reshape(-1, 28, 28, 1)
test_images = test_images.reshape(-1, 28, 28, 1)

# 모델 학습
model.fit(train_images, train_labels, epochs=50, batch_size=128, validation_split=0.3)

# 테스트 데이터로 모델 평가
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('테스트 정확도:', test_acc)


4422102/4422102 [==============================] - 0s 0us/step
Epoch 1/50
329/329 [==============================] - 13s 7ms/step - loss: 0.8504 - accuracy: 0.6943 - val_loss: 0.5153 - val_accuracy: 0.8055
Epoch 2/50
329/329 [==============================] - 2s 5ms/step - loss: 0.5037 - accuracy: 0.8193 - val_loss: 0.3925 - val_accuracy: 0.8589
Epoch 3/50
329/329 [==============================] - 2s 5ms/step - loss: 0.4111 - accuracy: 0.8540 - val_loss: 0.3497 - val_accuracy: 0.8723
Epoch 4/50
329/329 [==============================] - 2s 5ms/step - loss: 0.3612 - accuracy: 0.8738 - val_loss: 0.3251 - val_accuracy: 0.8836
Epoch 5/50
329/329 [==============================] - 2s 7ms/step - loss: 0.3314 - accuracy: 0.8845 - val_loss: 0.3451 - val_accuracy: 0.8726
Epoch 6/50
329/329 [==============================] - 2s 7ms/step - loss: 0.3074 - accuracy: 0.8909 - val_loss: 0.3077 - val_accuracy: 0.8892
Epoch 7/50
329/329 [==============================] - 2s 5ms/step - loss: 0.2907 - a

In [2]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 14, 14, 64)        640       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 7, 7, 64)         0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 4, 4, 64)          36928     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 2, 2, 64)         0         
 2D)                                                             
                                                                 
 flatten (Flatten)           (None, 256)               0         
                                                                 
 dense (Dense)               (None, 256)               6

첫 번째 Conv2D 레이어
- 입력 크기: 28x28x1
- 필터 크기: 3x3
- 필터 개수: 64
- 스트라이드: 2
- 패딩: "same"
- 파라미터 수: (3x3x1 + 1) x 64 = 640

첫 번째 MaxPooling2D 레이어:
- 풀링 크기: 2x2
- 파라미터 수: 0 (풀링 레이어는 학습되는 파라미터가 없음)

두 번째 Conv2D 레이어:
- 입력 크기: 14x14x64 (이전 레이어에서 풀링으로 인해 크기가 줄어듦)
- 필터 크기: 3x3
- 필터 개수: 64
- 스트라이드: 2
- 패딩: "same"
- 파라미터 수: (3x3x64 + 1) x 64 = 36928

두 번째 MaxPooling2D 레이어:
- 풀링 크기: 2x2
- 파라미터 수: 0 (풀링 레이어는 학습되는 파라미터가 없음)

총 CNN 파라미터 수: 640 + 0 + 36928 + 0 = 37568

따라서, 위 (3) 모델의 CNN 계층은 총 37,568개의 파라미터를 가지고 있습니다

(1) 차원축소는 차원의 저주로 인해 발생하는 몇 가지 문제점을 해소할 수 있습니다. 차원의 저주란, 고차원 공간에서 데이터 포인트들이 희소하게 분포하여 데이터 분석이 어렵고 비효율적이라는 현상을 의미합니다. 차원이 증가할수록 데이터 포인트 사이의 거리가 멀어지고, 데이터의 분포를 잘 표현하기 위해 필요한 데이터 양도 기하급수적으로 증가합니다.

차원축소는 고차원 데이터를 저차원으로 변환함으로써 이러한 문제를 해결할 수 있습니다. 저차원으로 변환된 데이터는 데이터의 분포를 더 잘 표현하고, 데이터 간의 거리가 더 가까워져 유사한 패턴이나 관계를 더 잘 파악할 수 있게 됩니다. 이를 통해 데이터 분석이 간편해지고, 예측 모델의 성능 향상이 기대됩니다.

(2) 차원 축소 과정에서 감수해야 하는 단점은 다음과 같습니다.
- 정보 손실: 차원 축소는 원래 데이터의 차원을 줄이기 때문에 일부 정보가 손실될 수 있습니다. 이는 원래 데이터의 특징이나 변동성을 완벽하게 보존하지 못하고 압축된 형태로 표현됨을 의미합니다.

- 해석의 어려움: 저차원으로 축소된 데이터는 원래 데이터보다 해석하기 어려울 수 있습니다. 축소된 차원은 원래 데이터의 특성을 추상화하고 합성한 형태이기 때문에, 이를 해석하고 이해하기가 어려울 수 있습니다.

- 계산 비용: 일부 차원 축소 알고리즘은 계산적으로 비용이 많이 들 수 있습니다. 특히, 대규모 데이터셋이나 고차원 데이터의 경우에는 계산 비용이 급증할 수 있습니다.

(3) autoencoder를 이용한 차원축소와 주성분 분석(PCA)을 통한 차원 축소의 차이점은 다음과 같습니다:

- 방법론: autoencoder는 신경망 기반의 비지도학습 방법으로, 입력 데이터를 잠재 공간으로 인코딩하고, 다시 재구성하는 과정을 통해 차원 축소를 수행합니다. 반면에 PCA는 선형 변환을 사용하여 데이터의 분산을 최대로 보존하는 저차원 표현을 찾는 방법입니다.
- 유연성: autoencoder는 비선형 변환을 통해 복잡한 데이터 구조를 학습할 수 있으므로, 비선형 관계를 갖는 데이터에 대해서도 잘 동작할 수 있습니다. 반면에 PCA는 선형 변환만을 사용하므로, 비선형 구조를 잘 반영하지 못할 수 있습니다.
- 지도/비지도 학습: autoencoder는 비지도학습 방법이므로, 입력 데이터에 대한 레이블 정보를 사용하지 않습니다. 반면에 PCA는 주로 비지도학습으로 사용되지만, 필요에 따라 입력 데이터의 레이블 정보를 활용하여 데이터 분산을 최대화하는 방향으로 변환을 수행할 수도 있습니다.

Machine learning과 deep learning은 인공지능 분야에서 주요한 개념이지만, 그들 간에는 몇 가지 중요한 차이점이 있습니다. 아래에서는 이러한 차이점을 상세히 설명하겠습니다.

1. 구조와 아키텍처:

- Machine learning: Machine learning은 데이터를 학습하고 패턴을 발견하기 위해 통계적 모델을 사용합니다. 주로 특성 추출, 모델 선택 및 하이퍼파라미터 조정과 같은 전처리 과정을 포함합니다. 전통적인 기계 학습 방법에는 회귀, 분류, 군집화 및 차원 축소와 같은 다양한 알고리즘이 포함됩니다.
- Deep learning: Deep learning은 인공신경망(artificial neural network)이라고도 불리는 계층적인 구조를 사용하여 패턴을 학습합니다. Deep learning은 데이터의 표현을 자동으로 학습하며, 복잡한 비선형 관계를 모델링할 수 있습니다. Deep learning은 다양한 은닉층(hidden layer)으로 구성된 신경망을 사용하며, 주로 이미지, 음성 및 자연어 처리와 같은 고수준의 작업에 사용됩니다.

2. 데이터 요구 사항:

- Machine learning: Machine learning은 상대적으로 적은 양의 데이터로도 좋은 결과를 얻을 수 있습니다. 데이터의 특성을 수동으로 추출해야 할 수도 있지만, 데이터의 양이 제한적인 경우에도 일반적으로 사용됩니다.
- Deep learning: Deep learning은 대규모의 데이터셋이 필요합니다. Deep learning 모델은 대용량의 데이터를 사용하여 표현력과 일반화 능력을 향상시키는 경향이 있습니다. 따라서 Deep learning은 대규모 데이터셋을 사용할 수 있는 경우에 가장 효과적입니다.

3. 특성 추출:

- Machine learning: Machine learning에서는 특성 추출(feature extraction)과 선택이 중요한 단계입니다. 전통적인 기계 학습에서는 도메인 전문가가 특성을 정의하고 추출하는 과정이 필요합니다. 이러한 특성은 일반적으로 데이터의 주요 속성을 나타내는 데 도움이 됩니다.
- Deep learning: Deep learning은 데이터로부터 자동으로 특성을 학습합니다. Deep learning 모델은 입력 데이터에서 특징을 추출하는 과정을 내부적으로 처리합니다. 이는 데이터에서 가장 유용한 특징을 학습하고 표현을 자동으로 학습함으로써 많은 특성 공학 단계를 제거합니다.

4. 계산 리소스:

- Machine learning: Machine learning 모델은 일반적으로 상대적으로 적은 계산 리소스를 요구합니다. 대부분의 기계 학습 알고리즘은 비교적 간단하고 계산적으로 경제적입니다.
- Deep learning: Deep learning은 많은 계산 리소스를 필요로 합니다. Deep learning 모델은 많은 수의 파라미터와 계산을 필요로 하기 때문에 고성능 하드웨어(예: GPU 또는 TPU)를 사용하여 훈련과 추론을 수행하는 것이 일반적입니다.

5. 활용 분야:

- Machine learning: Machine learning은 다양한 분야에서 사용됩니다. 예를 들어, 주가 예측, 이메일 스팸 필터링, 고객 세분화, 의료 진단, 추천 시스템 등에 적용될 수 있습니다.
- Deep learning: Deep learning은 주로 이미지 분류, 객체 탐지, 음성 인식, 자연어 처리, 기계 번역 등의 고수준 작업에 사용됩니다. 특히 컴퓨터 비전과 음성 처리 분야에서 많은 성과를 내고 있습니다.
이러한 차이점을 통해 Machine learning과 Deep learning은 서로 보완적인 개념으로 이해될 수 있습니다. Machine learning은 비교적 간단하고 데이터 요구 사항이 적은 경우에 유용하며, Deep learning은 대규모 데이터셋과 계산 리소스가 있는 경우에 더 나은 성능을 발휘합니다.